### PR TITLE
Handle installation when subscription expired

### PR DESCRIPTION
**Pull request by jesusbv**
- Check system registration and subscription status
- Remove repositories and install unrestricted repo
  packages if subscription is not active

### CODE REVIEW

File: main/ahb_for_sles.go [hunk 1]
```diff
@@ -97,13 +98,49 @@ func parseCfg(filename string) (map[string]map[string]string, error) {
 	return ini, nil
 }
 
-func _isRegistered() (bool, error) {
-	services, err := filepath.Glob("/etc/zypp/services.d/*.service")
-	if err != nil {
-		fmt.Fprintln(os.Stderr, err)
-		return false, err
+func _getSUSEConnectStatus() (bool, bool, error) {
```
Comment (simratsatia) [inline main/ahb_for_sles.go:101]: @jesusbv, please ensure that all error events are getting logged in the telemetry.
Eg: ext.ExtensionEvents.LogErrorEvent(ENABLE_EVENT,fmt.Sprintf(OPERATION_FAILURE_MSG, ENABLE_EVENT, err.Error()))

Using object ext.ExtensionEvents, you should be able to send important telemetry to Azure Kusto. This is to ensure much ease in debugging any future errors that customers might face and correspondingly witness how the extension is performing.

File: main/ahb_for_sles.go [hunk 2]
```diff
@@ -97,13 +98,49 @@ func parseCfg(filename string) (map[string]map[string]string, error) {
...
+	if strings.ToLower(status) == "registered" && strings.ToLower(subscriptionStatus) == "active" {
+		// system is registered with an active subscription
+		// check if services are present
+		services, err := filepath.Glob("/etc/zypp/services.d/*.service")
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return true, true, err
+		}
+		if len(services) == 0 {
+			errorMessage := "There are no services. Please, re register the system"
```
Comment (rjschwei) [inline main/ahb_for_sles.go:120]: The system is already registered, if we tell the user to re-register than 2 subscriptions will be counted against one system. That is bad advice at best. We need to tell the user the correct action to take, or restore the services. The service restore is fairly straight forward:

```
SUSEConnect --list-extensions
```
Then loop through every product that has already been `(Activated)` or said differently for every line in the output that has "Deactivate with:" take the presented `SUSEConnect` command and drop the `-d`

File: main/ahb_for_sles.go [hunk 3]
```diff
@@ -97,13 +98,49 @@ func parseCfg(filename string) (map[string]map[string]string, error) {
...
+			return true, true, err
+		}
+		// check if repos are present
+		repos, err := filepath.Glob("/etc/zypp/repos.d/*.repo")
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return true, true, err
+		}
+		if len(repos) == 0 {
+			errorMessage := "There are no repositores. Please, re register the system"
```
Comment (rjschwei) [inline main/ahb_for_sles.go:132]: This is also not correct, if there is a service, which we have if the code got this far, then the repo definitions do not matter as a `zypper ref -f` will restore the repos. That's the whole point of the service ;) Repository Information Service.

**Review by rjschwei — CHANGES_REQUESTED** [review]

File: main/ahb_for_sles.go (hunk 2 above)
Comment (jesusbv) [reply main/ahb_for_sles.go]: Done, also activate Public Cloud module in case it was not active before

File: main/ahb_for_sles.go (hunk 3 above)
Comment (jesusbv) [reply main/ahb_for_sles.go]: Done, used he verbose option of `refs` option instead of `ref`

File: main/ahb_for_sles.go [hunk 4]
```diff
@@ -97,61 +98,88 @@ func parseCfg(filename string) (map[string]map[string]string, error) {
 	return ini, nil
 }
 
-func _isRegistered() (bool, error) {
-	services, err := filepath.Glob("/etc/zypp/services.d/*.service")
-	if err != nil {
-		fmt.Fprintln(os.Stderr, err)
-		return false, err
+func _getSUSEConnectStatus() (bool, bool, error) {
```
Comment (rjschwei) [inline main/ahb_for_sles.go:101]: The function name is suspect. `_getSUSEConnectStatus` should return a status, based on name and should not have any side effects. But the side effect of this call is that, depending on the state of the system, repositories may get enabled. We should strive for side effect free implementations.

File: main/ahb_for_sles.go [hunk 5]
```diff
@@ -97,61 +98,88 @@ func parseCfg(filename string) (map[string]map[string]string, error) {
...
+	commandOutput, error := RunShellCommand(0, "SUSEConnect", "-s")
+	if error != nil {
+		fmt.Println(error)
+		return false, false, error
+	}
+	var suseConnectStatus []map[string]interface{}
+	json.Unmarshal([]byte(commandOutput), &suseConnectStatus)
+	subscriptionStatus := fmt.Sprintf("%v", suseConnectStatus[0]["subscription_status"])
+	status := fmt.Sprintf("%v", suseConnectStatus[0]["status"])
+	if strings.ToLower(status) == "registered" && strings.ToLower(subscriptionStatus) == "active" {
```
Comment (rjschwei) [inline main/ahb_for_sles.go:111]: This is where a status check would end, the rest is side effect coding.

File: main/ahb_for_sles.go [hunk 6]
```diff
@@ -97,61 +98,88 @@ func parseCfg(filename string) (map[string]map[string]string, error) {
...
+							fmt.Fprintln(os.Stderr, err)
+							return true, true, err
+						}
+					}
+				}
+			}
+			return true, true, err
+		}
+		// check if repos are present
+		repos, err := filepath.Glob("/etc/zypp/repos.d/*.repo")
```
Comment (rjschwei) [inline main/ahb_for_sles.go:145]: If we have services then `zypper` will set up the repos. For example you can `rm -rf /etc/zypp/repos.d/*.repo` on a system and as long as the service is set up can run `zypper up` and the repos will be restored. I am not certain why we need the extra check for repositories. I think conceptually there is some confusion about this use case implementation.

**Review by rjschwei — CHANGES_REQUESTED** [review]

**Comment by rjschwei** [comment]
Code independent comment because I think there is some confusion. Process steps in pseudo code:

- check registration status (SUSEConnect -s)
- if registered and active
  - check_has_services (not part of the "check registration status" implementation)
  - if no services
    - restore services
  - nothing to do
  - check has public cloud module
    - if has public cloud module -> nothing to do
    - else enable the public cloud module
- else
  - setup unrestricted repo

File: main/ahb_for_sles.go [hunk 7]
```diff
@@ -97,61 +98,88 @@ func parseCfg(filename string) (map[string]map[string]string, error) {
...
+			extensionsOutput, err := RunShellCommand(0, "SUSEConnect", "--list-extensions")
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err)
+				return true, true, err
+			}
+			extensionsList := strings.Split(extensionsOutput, "\n")
+			for _, extension := range extensionsList {
+				if strings.Contains(extension, "Deactivate with") {
+					// activate whatever it was active AND
+					// Public Cloud module
```
[the remark this answers is not in the document]
Comment (jesusbv) [reply main/ahb_for_sles.go:129]: I removed the code but missed the comment, will update the comment

**Review by rjschwei — CHANGES_REQUESTED** [review]

File: main/ahb_for_sles.go [hunk 8]
```diff
@@ -240,79 +233,203 @@ func _installPackages(ahbInfo AHBInfo) error {
...
+		fmt.Println("Error getting repositories from '/etc/zypp/repos.d'")
 		return err
 	}
+	for _, repo := range repos {
+		repoFile, err := os.Open(repo)
 
-	if isRegistered {
-		hasPubCloudMod, distro, arch, errGlob := _hasPubCloudMod(ahbInfo.PublicCloudService)
+		if err != nil {
+			fmt.Println(err)
```
[the remark this answers is not in the document]
Comment (jesusbv) [reply main/ahb_for_sles.go]: Done

**Review by rjschwei — APPROVED** [review]

**Pull request merged**